Handle updates removing remaining finalizers on deleted objects

liggitt opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Avoids making all finalizer implementers do a separate delete after removing their finalizer

**Which issue(s) this PR fixes**:
Fixes #77944

**Does this PR introduce a user-facing change?**:
```release-note
Updates that remove remaining `metadata.finalizers` from  an object that is pending deletion (non-nil metadata.deletionTimestamp) and has no graceful deletion pending (nil or 0 metadata.deletionGracePeriodSeconds) now results in immediate deletion of the object.
```

/sig api-machinery
/cc @caesarxuchao

liggitt committed:
Handle updates removing remaining finalizers on deleted objects

caesarxuchao commented:
The approach lgtm.

Please add a unit test to ensure the namespace delete behavior is correct.

We need to update namespace controller logic, these lines should be redundant now: https://github.com/kubernetes/kubernetes/blob/aaec77a94b67878ca1bdd884f2778f4388d203f2/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L157-L159

https://github.com/kubernetes/kubernetes/blob/aaec77a94b67878ca1bdd884f2778f4388d203f2/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go#L158-L159

liggitt commented:
> Please add a unit test to ensure the namespace delete behavior is correct.

done, added positive and negative unit tests for spec and metadata finalizers on namespace update and finalize storage

> We need to update namespace controller and crd finalizer logic, these lines should be redundant now

added TODOs to remove redundant deletes in the next release

liggitt commented:
/retest

caesarxuchao commented:
/lgtm

liggitt commented:
/priority important-soon
